src/dataflow-types/src/client.rs

```diff
@@ -964,3 +1122,248 @@ pub mod process_local {
         }
     }
 }
+
+/// A client to a remote dataflow server.
+pub mod tcp {
+    use async_trait::async_trait;
+    use futures::sink::SinkExt;
+    use futures::stream::StreamExt;
+    use tokio::io::{AsyncRead, AsyncWrite};
+    use tokio::net::{TcpStream, ToSocketAddrs};
+    use tokio_serde::formats::Bincode;
+    use tokio_util::codec::LengthDelimitedCodec;
+
+    use crate::client::{Command, GenericClient, Response};
+
+    /// A client to a remote dataflow server.
+    #[derive(Debug)]
+    pub struct TcpClient<T> {
+        connection: FramedClient<TcpStream, T>,
+    }
+
+    impl<T> TcpClient<T> {
+        /// Connects a remote client to the specified remote dataflow server.
+        pub async fn connect(addr: impl ToSocketAddrs) -> Result<TcpClient<T>, anyhow::Error> {
+            let connection = framed_client(TcpStream::connect(addr).await?);
+            Ok(Self { connection })
+        }
+    }
+
+    #[async_trait]
+    impl<T> GenericClient<Command<T>, Response<T>> for TcpClient<T>
+    where
+        T: timely::progress::Timestamp + Copy + Unpin,
+    {
+        async fn send(&mut self, cmd: Command<T>) -> Result<(), anyhow::Error> {
+            // TODO: something better than panicking.
+            self.connection.send(cmd).await.map_err(|err| err.into())
+        }
+
+        async fn recv(&mut self) -> Option<Response<T>> {
+            // TODO: something better than panicking.
+            self.connection
+                .next()
+                .await
+                .map(|x| x.expect("connection to dataflow server broken"))
+        }
+    }
+
+    /// A framed connection to a dataflowd server.
+    pub type Framed<C, T, U> = tokio_serde::Framed<
+        tokio_util::codec::Framed<C, LengthDelimitedCodec>,
```

Comment:
My eyes glazed over reading this. Is there a good way to document what the type parameters mean? Alternately, longer names to use for them?